### PR TITLE
SimplicialComplex test autogen

### PR DESCRIPTION
**Pull request by zlyfunction**
add autogenerated tests for SimplicalComplex tests for 2d

### CODE REVIEW

File: tests/auto_generated/SimplicialComplex_test_autogen.py [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+        return mesh.link
+    elif test_function_name == "simplex_with_boundary":
+        return mesh.boundary
+    elif test_function_name == "top_coface_simplex":
+        return mesh.top_coface_simplex
+    else:
+        raise Exception("Function name not recognized")
+    
+def write_tests(file, path_to_data, mesh_name, n_samples, test_function_names):
+    mesh = load_obj(path_to_data + mesh_name)
```
Comment (mtao) [inline tests/auto_generated/SimplicialComplex_test_autogen.py:38]: there needs to also be functionality for randomly deleting faces and and checkign that the result of those deletions is a manifold mesh

File: tests/auto_generated/SimplicialComplex_test_autogen.py [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+                for i in range(len(simplex)):
+                    file.write("    CHECK(sc_f[{}][{}] == {});\n".format(sc.index(simplex), i, simplex[i]))
+            file.write("\n")
+        
+        file.write("}\n")
+
+
+
+# Parameters
+path = '../../data/'
```
Comment (mtao) [inline tests/auto_generated/SimplicialComplex_test_autogen.py:115]: this path might not be canonical so it should be an argument for the script with some instructions on what should be found in the path (such as being some subfolder of WMTK_DATA_DIR?)

**Review by mtao — CHANGES_REQUESTED** [review]

File: tests/auto_generated/SimplicialComplex.py [hunk 3]
```diff
@@ -0,0 +1,112 @@
+from itertools import chain, combinations
+import igl
+import numpy as np
+class Mesh:
+    def __init__(self):
+        self.vertices = []    # List of vertices
+        self.tetrahedra = []  # List of tetrahedra
```
[the remark this answers is not in the document]
Comment (zlyfunction) [reply tests/auto_generated/SimplicialComplex.py:7]: tets will alway be the top simplex, so there will be no redundant tets in init.

File: tests/auto_generated/test_simplicial_complex_2d.cpp [hunk 4]
```diff
@@ -0,0 +1,3286 @@
...
+    constexpr PrimitiveType PV = PrimitiveType::Vertex;
+    constexpr PrimitiveType PE = PrimitiveType::Edge;
+    constexpr PrimitiveType PF = PrimitiveType::Face;
+    constexpr PrimitiveType PT = PrimitiveType::Tetrahedron;
+    TEST_CASE("open_star_circle", "[simplicial_complex][open_star][2D][auto_generated]")
+{
+    RowVectors3d V;
+    RowVectors3l F;
+    std::string name = "/circle.obj";
+    std::string path;
```
Comment (daniel-zint) [inline tests/auto_generated/test_simplicial_complex_2d.cpp:20]: use `std::filesystem::path`

File: tests/tools/DEBUG_TriMesh.cpp [hunk 5]
```diff
@@ -174,4 +173,41 @@ auto DEBUG_TriMesh::get_tmoe(const Tuple& t, Accessor<long>& hash_accessor)
 {
     return TriMeshOperationExecutor(*this, t, hash_accessor);
 }
+
+std::vector<std::vector<long>> DEBUG_TriMesh::get_sorted_simplicial_complex(const std::vector<Simplex> &sc) const
+{
+    std::vector<std::vector<long>> ret;
+    for (auto s : sc)
```
Comment (daniel-zint) [inline tests/tools/DEBUG_TriMesh.cpp:180]: `for(const auto& s : sc)`

File: tests/tools/DEBUG_TriMesh.cpp [hunk 6]
```diff
@@ -174,4 +173,41 @@ auto DEBUG_TriMesh::get_tmoe(const Tuple& t, Accessor<long>& hash_accessor)
 {
     return TriMeshOperationExecutor(*this, t, hash_accessor);
 }
+
+std::vector<std::vector<long>> DEBUG_TriMesh::get_sorted_simplicial_complex(const std::vector<Simplex> &sc) const
```
Comment (daniel-zint) [inline tests/tools/DEBUG_TriMesh.cpp:177]: The name of this function is not very intuitive. Please, either provide a better name or add a comment describing what this function does.

File: tests/auto_generated/test_simplicial_complex_2d.cpp [hunk 7]
```diff
@@ -0,0 +1,3286 @@
...
+    path.append(WMTK_DATA_DIR);
+    path.append(name);
+    igl::read_triangle_mesh(path, V, F);
+    tests::DEBUG_TriMesh m;
+    m.initialize(F);
+
+    Tuple t;
+    std::vector<std::vector<long>> sc_v, sc_e, sc_f;
+    t = m.tuple_from_id(PV, 5058);
+    sc_v = m.get_sorted_simplicial_complex(SimplicialComplex::open_star(m, Simplex(PV, t)).get_simplex_vector());
```
Comment (daniel-zint) [inline tests/auto_generated/test_simplicial_complex_2d.cpp:30]: I'd prefer `const auto sc_v = ...` here.

File: tests/auto_generated/test_simplicial_complex_2d_hard.cpp [hunk 8]
```diff
@@ -0,0 +1,4528 @@
...
+
+    constexpr PrimitiveType PV = PrimitiveType::Vertex;
+    constexpr PrimitiveType PE = PrimitiveType::Edge;
+    constexpr PrimitiveType PF = PrimitiveType::Face;
+    constexpr PrimitiveType PT = PrimitiveType::Tetrahedron;
+    TEST_CASE("open_star_circle_0", "[simplicial_complex][open_star][2D][auto_generated]")
+{
+    RowVectors3d V;
+    RowVectors3l F;
+    std::string name = "/circle_0.7.obj";
```
Comment (daniel-zint) [inline tests/auto_generated/test_simplicial_complex_2d_hard.cpp:19]: I don't think that file is in the data repo

**Review by daniel-zint — COMMENTED** [review]
I need to double check if all cases are covered by the tests. Otherwise, I added some minor comments.

File: tests/auto_generated/SimplicialComplex_test_autogen.py [hunk 9]
```diff
@@ -0,0 +1,114 @@
...
+mesh_names = ['circle.obj', 'bunny.obj','blub.obj']
+output_file_name = 'test_simplicial_complex_2d.cpp'
+test_function_names = ['open_star', 'closed_star', 'link', 'simplex_with_boundary', 'top_coface_simplex']
```
Comment (daniel-zint) [inline tests/auto_generated/SimplicialComplex_test_autogen.py:106]: missing functions that should be tested
* `boundary`
* `simplices_w_boundary_intersect`
* `link_cond`
* `vertex_one_ring`
* `k_ring`

**Review by daniel-zint — COMMENTED** [review]
There are some function tests missing.

File: tests/auto_generated/test_simplicial_complex_2d_hard.cpp (hunk 8 above)
Comment (mtao) [reply tests/auto_generated/test_simplicial_complex_2d_hard.cpp:19]: this belongs in the data repo - separate your code for auto-generating meshes and for validating meshes into two parts:
* generation goes in teh data repo + some generated meshes
* test generation lives in `tests/auto_generated/`

File: tests/auto_generated/SimplicialComplex_test_autogen.py [hunk 10]
```diff
@@ -0,0 +1,114 @@
...
+    # also sample some simplex on the boudnary of the mesh
+    v, _, _, f, _, _ = igl.read_obj(path_to_data + mesh_name)
+    bd_loop = igl.boundary_loop(f)
+    if len(bd_loop) > 0:
+        n_bd_samples = min(n_samples, len(bd_loop))
+        indices = list(range(len(bd_loop) - 1))
+        for _ in range(n_bd_samples):
+            index = random.choice(indices)  # Choose a random index
+            v_samples.append((bd_loop[index],))
+            e_samples.append(tuple(sorted([bd_loop[index], bd_loop[(index + 1) % len(bd_loop)]])))
```
Comment (daniel-zint) [inline tests/auto_generated/SimplicialComplex_test_autogen.py:41]: You should also test edges that are incident to the boundary, i.e. one vertex on the boundary and one on the interior. The same holds for faces.

Comment (daniel-zint) [reply tests/auto_generated/SimplicialComplex_test_autogen.py:41]: Also, there might be edges that are interior but both vertices are on the boundary. Any possible combination should be tested.

**Comment by mtao** [comment]
also, auto_generated should just be our scripts folder - which already handles our other autogen code
